tfsdk/serve.go

```diff
@@ -216,15 +216,73 @@ func (s *server) GetProviderSchema(ctx context.Context, _ *tfprotov6.GetProvider
 }
 
 func (s *server) ValidateProviderConfig(ctx context.Context, req *tfprotov6.ValidateProviderConfigRequest) (*tfprotov6.ValidateProviderConfigResponse, error) {
-	// uncomment when we implement this function
-	// ctx = s.registerContext(ctx)
-
-	// We don't actually do anything as part of this. In theory, we could
-	// validate the configuration for the provider block? Need to check in
-	// again with the core team about the goal of this RPC.
-	return &tfprotov6.ValidateProviderConfigResponse{
+	ctx = s.registerContext(ctx)
+	resp := &tfprotov6.ValidateProviderConfigResponse{
+		// This RPC allows a modified configuration to be returned. This was
+		// previously used to allow a "required" provider attribute (as defined
+		// by a schema) to still be "optional" with a default value, typically
+		// through an environment variable. Other tooling based on the provider
+		// schema information could not determine this implementation detail.
+		// To ensure accuracy going forward, this implementation is opinionated
+		// towards accurate provider schema definitions and optional values
+		// can be filled in or return errors during ConfigureProvider().
 		PreparedConfig: req.Config,
-	}, nil
+	}
+
+	schema, diags := s.p.GetSchema(ctx)
+
+	if diags != nil {
+		resp.Diagnostics = append(resp.Diagnostics, diags...)
+
+		if diagsHasErrors(resp.Diagnostics) {
+			return resp, nil
+		}
+	}
+
+	config, err := req.Config.Unmarshal(schema.TerraformType(ctx))
+
+	if err != nil {
+		resp.Diagnostics = append(resp.Diagnostics, &tfprotov6.Diagnostic{
+			Severity: tfprotov6.DiagnosticSeverityError,
+			Summary:  "Error parsing config",
+			Detail:   "The provider had a problem parsing the config. Report this to the provider developer:\n\n" + err.Error(),
+		})
+
+		return resp, nil
+	}
+
+	vpcReq := ValidateProviderConfigRequest{
+		Config: Config{
+			Raw:    config,
+			Schema: schema,
+		},
+	}
+
+	if provider, ok := s.p.(ProviderWithConfigValidators); ok {
+		for _, configValidator := range provider.ConfigValidators(ctx) {
+			vpcRes := &ValidateProviderConfigResponse{}
+
+			configValidator.Validate(ctx, vpcReq, vpcRes)
+
+			if vpcRes.Diagnostics != nil {
+				resp.Diagnostics = append(resp.Diagnostics, vpcRes.Diagnostics...)
+			}
+		}
+	}
+
+	if provider, ok := s.p.(ProviderWithValidateConfig); ok {
+		vpcRes := &ValidateProviderConfigResponse{}
```

Comment:
Do we want to insert any existing diagnostics into the response before passing it on?

Reply:
This is an interesting proposition! I see the value in it -- essentially being able to change or remove diagnostics with these validation methods. I'm a little torn on the implementation however, as it feels weird to pre-populate part of a response and have developers of these methods understand the semantics of those potentially pre-populated responses. It _feels_ like to support this we may want to potentially pass in a "prior diagnostics" request field and either allow developers to return updated/removed diagnostics in the existing response `Diagnostics` field (although it seems like they could forget to passthrough existing ones) or provide separate fields in the response for this special handling. I'm a little worried that this would be different than other diagnostics handling, which is currently append-only everywhere else.

Reply:
What I'm trying to get at is I _think_ everywhere else we populate the response object we pass to the provider with the diagnostics we already have (if any), and they append to it, and we replace the gRPC response diagnostics with what comes back. ... I think.

I don't know that it matters, I'm mostly just arguing for consistency.

Reply:
Updated in faa5afb1c9f191bf4533743c8848966713337a69 and ee6ea5de2d21dafd19ecdc0e5bf2e6c50a3633b8 for consistency with other functions.